File: packages/react-native/ReactCommon/react/runtime/hermes/HermesInstance.cpp
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include "HermesInstance.h"

#include <hermes/inspector-modern/chrome/HermesRuntimeTargetDelegate.h>
#include <jsi/jsilib.h>
#include <jsinspector-modern/InspectorFlags.h>
#include <react/featureflags/ReactNativeFeatureFlags.h>

using namespace facebook::hermes;
using namespace facebook::jsi;

namespace facebook::react {

class HermesJSRuntime : public JSRuntime {
 public:
  HermesJSRuntime(std::unique_ptr<HermesRuntime> runtime)
      : runtime_(std::move(runtime)) {}

  jsi::Runtime& getRuntime() noexcept override {
    return *runtime_;
  }

  jsinspector_modern::RuntimeTargetDelegate& getRuntimeTargetDelegate()
      override {
    if (!targetDelegate_) {
      targetDelegate_.emplace(runtime_);
    }
    return *targetDelegate_;
  }

  void unstable_initializeOnJsThread() override {
    runtime_->registerForProfiling();
  }

 private:
  std::shared_ptr<HermesRuntime> runtime_;
  std::optional<jsinspector_modern::HermesRuntimeTargetDelegate>
      targetDelegate_;
};

std::unique_ptr<JSRuntime> HermesInstance::createJSRuntime(
    std::shared_ptr<::hermes::vm::CrashManager> crashManager,
    std::shared_ptr<MessageQueueThread> msgQueueThread,
    bool allocInOldGenBeforeTTI) noexcept {
  assert(msgQueueThread != nullptr);

  auto gcConfig = ::hermes::vm::GCConfig::Builder()
                      // Default to 3GB
                      .withMaxHeapSize(3072 << 20)
                      .withName("RNBridgeless");

  if (allocInOldGenBeforeTTI) {
    // For the next two arguments: avoid GC before TTI
    // by initializing the runtime to allocate directly
    // in the old generation, but revert to normal
    // operation when we reach the (first) TTI point.
    gcConfig.withAllocInYoung(false).withRevertToYGAtTTI(true);
  }

  ::hermes::vm::RuntimeConfig::Builder runtimeConfigBuilder =
      ::hermes::vm::RuntimeConfig::Builder()
          .withGCConfig(gcConfig.build())
          .withEnableSampleProfiling(true)
          .withMicrotaskQueue(
              ReactNativeFeatureFlags::enableBridgelessArchitecture());

  if (crashManager) {
    runtimeConfigBuilder.withCrashMgr(crashManager);
  }

  std::unique_ptr<HermesRuntime> hermesRuntime =
      hermes::makeHermesRuntime(runtimeConfigBuilder.build());

  auto errorPrototype = hermesRuntime->global()
                            .getPropertyAsObject(*hermesRuntime, "Error")
                            .getPropertyAsObject(*hermesRuntime, "prototype");
  errorPrototype.setProperty(*hermesRuntime, "jsEngine", "hermes");

  (void)msgQueueThread;

  return std::make_unique<HermesJSRuntime>(std::move(hermesRuntime));
}

} // namespace facebook::react

File: packages/react-native/temp/package/ReactAndroid/src/main/java/com/facebook/react/runtime/hermes/HermesInstance.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.runtime.hermes

import com.facebook.jni.HybridData
import com.facebook.jni.annotations.DoNotStrip
import com.facebook.react.fabric.ReactNativeConfig
import com.facebook.react.runtime.JSRuntimeFactory
import com.facebook.soloader.SoLoader

public class HermesInstance(
    reactNativeConfig: ReactNativeConfig?,
    allocInOldGenBeforeTTI: Boolean
) : JSRuntimeFactory(initHybrid(reactNativeConfig as Any?, allocInOldGenBeforeTTI)) {

  public constructor() : this(null, false)

  public companion object {
    @JvmStatic
    @DoNotStrip
    protected external fun initHybrid(
        reactNativeConfig: Any?,
        allocInOldGenBeforeTTI: Boolean
    ): HybridData

    init {
      SoLoader.loadLibrary("hermesinstancejni")
    }
  }
}
